numpy/core/src/multiarray/mapping.c

```diff
@@ -912,38 +912,53 @@ get_view_from_index(PyArrayObject *self, PyArrayObject **view,
                     npy_index_info *indices, int index_num, int ensure_array) {
     npy_intp new_strides[NPY_MAXDIMS];
     npy_intp new_shape[NPY_MAXDIMS];
+    npy_intp *self_dims = PyArray_DIMS(self);
     int i, j;
     int new_dim = 0;
     int orig_dim = 0;
     char *data_ptr = PyArray_BYTES(self);
+    int is_empty = 0;
 
     /* for slice parsing */
     npy_intp start, stop, step, n_steps;
 
+    /*
+     * prevent ptr + offset going beyond the end of the memory on empty arrays,
+     * which would be undefined behaviour according to the C99 spec paragraph
+     * 6.5.6 http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf
+     */
+    for (i=0; i < PyArray_NDIM(self); i++) {
+        if (self_dims[i] == 0) {
+            is_empty = 1;
+        }
+    }
+
```

Comment:
Wouldn't it be easier just to construct the strides on an empty array so that this doesn't happen?

Reply:
Also safer, since we probably perform this type of addition all over the place.

Reply:
strides seem to be mutable with little restrictions on their values. Dimensions less so.
```
a = np.zeros((1, 1, 0))
a.strides = (10000, 32, 12)  # passes
a.shape = (1, 1, 1)  #errors
```

Reply:
If strides are really that mutable, then the users can already invoke undefined behavior anyway.

Reply:
(This only applies to `a.size == 0` strides, otherwise there are checks in place to make sure `a.strides` does not allow access outside `a.data`). I am less concerned about undefined behaviour. I think 39e1233  is an improvement since it doesn't make sense for a view of a 0-sized array to point to different memory. I think this is (1) from the comment above

Reply:
Other places we use `PyArray_NewFromDescr{,AndBase}` with a possibly modified data pointer, all others call with `NULL` or with a `PyArray_DATA()` result: 
- `VOID_getitem`, `VOID_setitem` (`ip + offset` for structured dtypes),
- `arr_unravel_index` (seems safe),
- `PyArray_FromInterface` (the use of `offset` here looks suspicious, there are no checks that it is valid),
- `PyArray_FromBuffer` (checks validity), 
- `_get_part` (the implementation of `a.real` or `a.imag`, calculated internally and seems valid)
- `PyArray_Diagonal` (calculated internally, seems valid)
- `PyArray_Nonzero` (safe, the offset will always be 0 since `nonzero` is 0)
- `iter_array`, `npyiter_seq_item`, `npyiter_seq_ass_item` and `NpyIter_GetIterView` (too complicated to to dive into for this comment)
- `ufunc_at` (too complidated to dive into for this comment)

Additionally, the following API functions accept a data pointer `PArray_New`, `PyArray_CastScalarToCtypes`, anything goes with these.

So in conclusion, the only known dangerous one seems to be `offset` in `__array_interface__`, which we should probably have deprecated instead of fixed in gh-13036.